js/adapt-stateful-session.js

```diff
@@ -1,227 +1,259 @@
+          document.addEventListener;
+      if (shouldCommitOnVisibilityChange) {
+        document.addEventListener('visibilitychange', this.onVisibilityChange);
+      }
+      $(window).on('beforeunload unload', this.endSession);
+    }
 
-    removeEventListeners: function () {
-      $(window).off('beforeunload unload', this._onWindowUnload);
-      this.stopListening();
-    },
+    saveSessionState() {
+      const courseState = SCORMSuspendData.serialize([
+        Boolean(Adapt.course.get('_isComplete')),
+        Boolean(Adapt.course.get('_isAssessmentPassed'))
+      ]);
+      const componentStates = (this._shouldStoreResponses === true) ?
+        this._componentSerializer.serialize() :
+        '';
+      const sessionPairs = {
+        'c': courseState,
+        'q': componentStates
+      };
+      this.printCompletionInformation();
+      Adapt.offlineStorage.set(sessionPairs);
+    }
 
-    reattachEventListeners: function() {
+    printCompletionInformation() {
+      const courseComplete = Boolean(Adapt.course.get('_isComplete'));
+      const assessmentPassed = Boolean(Adapt.course.get('_isAssessmentPassed'));
+      const trackingIdModels = Adapt.data.filter(model => model.get('_type') === this._trackingIdType && model.has('_trackingId'));
+      const trackingIds = trackingIdModels.map(model => model.get('_trackingId'));
+      if (!trackingIds.length) {
+        Adapt.log.info(`course._isComplete: ${courseComplete}, course._isAssessmentPassed: ${assessmentPassed}, ${this._trackingIdType} completion: no tracking ids found`);
+        return;
+      }
+      const max = _.max(trackingIds);
+      const completion = trackingIdModels.reduce((completion, model) => {
+        const trackingId = model.get('_trackingId');
+        completion[trackingId] = model.get('_isComplete') ? '1' : '0';
+        return completion;
+      }, (new Array(max + 1)).join('-').split('')).join('');
+      Adapt.log.info(`course._isComplete: ${courseComplete}, course._isAssessmentPassed: ${assessmentPassed}, ${this._trackingIdType} completion: ${completion}`);
+    }
+
+    onLanguageChanged() {
+      // when the user switches language, we need to:
+      // - reattach the event listeners as the language change triggers a reload of
+      //   the json, which will create brand new collections
+      // - get and save a fresh copy of the session state. as the json has been reloaded,
+      //   the blocks completion data will be reset (the user is warned that this will
+      //   happen by the language picker extension)
+      // - check to see if the config requires that the lesson_status be reset to
+      //   'incomplete'
+      const config = Adapt.spoor.config;
       this.removeEventListeners();
       this.setupEventListeners();
-    },
-
-    onBlockComplete: function(block) {
       this.saveSessionState();
-    },
+      if (config._reporting && config._reporting._resetStatusOnLanguageChange === true) {
```

Comment:
Does `config` have to be checked for a truthy around here?

Reply:
done in https://github.com/adaptlearning/adapt-contrib-spoor/pull/197/commits/6fe523b804c56521147c4358729dd9fcc2c6fb11